core/src/core/sumtree.rs

```diff
@@ -0,0 +1,922 @@
+        }
+    }
+
+    /// Accessor for the tree's root
+    pub fn root_sum(&self) -> Option<(Hash, T::Sum)> {
+        self.root.as_ref().map(|node| node.root_sum())
+    }
+
+    fn insert_right_of(mut old: Node<T>, new: Node<T>) -> Node<T> {
+        assert!(old.depth >= new.depth);
+
+        // If we are inserting next to a full node, make a parent. If we're
+        // inserting a tree of equal depth then we get a full node, otherwise
+        // we get a partial node. Leaves and pruned data both count as full
+        // nodes.
+        if old.full == SubtreeFull::Yes {
+            let parent_depth = old.depth + 1;
+            let parent_sum = old.sum() + new.sum();
+            let parent_hash = (parent_depth, &parent_sum, old.hash, new.hash).hash();
+            let parent_full = if old.depth == new.depth {
+                SubtreeFull::Yes
+            } else {
+                SubtreeFull::No
+            };
+            let parent_data = NodeData::Internal {
+                lchild: Box::new(old),
+                rchild: Box::new(new),
+                sum: parent_sum,
+            };
+
+            Node {
+                full: parent_full,
+                data: parent_data,
+                hash: parent_hash,
+                depth: parent_depth
+            }
+        // If we are inserting next to a partial node, we should actually be
+        // inserting under the node, so we recurse. The right child of a partial
+        // node is always another partial node or a leaf.
+        } else {
+            if let NodeData::Internal{ ref lchild, ref mut rchild, ref mut sum } = old.data {
+                // Recurse
+                let dummy_child = Node { full: SubtreeFull::Yes, data: NodeData::Pruned(sum.clone()), hash: old.hash, depth: 0 };
+                let moved_rchild = mem::replace(&mut **rchild, dummy_child);
+                mem::replace(&mut **rchild, SumTree::insert_right_of(moved_rchild, new));
+                // Update this node's states to reflect the new right child
+                if rchild.full == SubtreeFull::Yes && rchild.depth == old.depth - 1 {
+                    old.full = rchild.full;
+                }
+                *sum = lchild.sum() + rchild.sum();
+                old.hash = (old.depth, &*sum, lchild.hash, rchild.hash).hash();
+            } else {
+                unreachable!()
+            }
+            old
+        }
+    }
+
+    /// Accessor for number of elements in the tree, not including pruned ones
+    pub fn len(&self) -> usize {
+        self.index.len()
+    }
+
+    /// Accessor for number of elements in the tree, including pruned ones
+    pub fn unpruned_len(&self) -> usize {
+        match self.root {
+            None => 0,
+            Some(ref node) => node.n_children()
+        }
+    }
+
+    /// Add an element to the tree. Returns true if the element was added,
+    /// false if it already existed in the tree.
+    pub fn push(&mut self, elem: T) -> bool {
```

Comment:
This method signature could stay the same but elem wouldn't be stored, just hashed and summed.